### PR TITLE
M3SA - Multi-Meta-Model Simulation Analyzer

### PR DESCRIPTION
**Pull request by Radu-Nicolae**
## Summary

M3SA - Multi-Meta-Model Simulation Analysis. Full functionality implemented and integrated into OpenDC.

## Multi-Model
- We design a novel simulation tool, able to leverage simulation data of multiple models into the same tool.
- The Multi-Model has extensive customizability for computation and plotting.

## Meta-Model
- We create the Meta-Model, which embeds functionality from the Multi-Model.
- The Meta-Model aggregates other models simulations, and is capable of performing meta-predictions.
- In plain English, the Meta-Model predicts using other models' predictions, merged using various functions (i.e., meta-functions), such as mean or median.

## M3SA
- We propose M3SA, a tool for ICT simulation analysis, using multi-model techniques. M3SA provides solutions to improve the performance of existing simulations of the Information and Communication Technology infrastructure.
- We release M3SA as free and open-source software for the community to use. M3SA follows modern, state-of-the-art engineering processes in the industry and produces reproducible results.
- M3SA integrated both the Multi-Model and the Meta-Model.
- Input: JSON interface
- Fully documented code
- Fully documented input and JSON schema validator
- M3SA can be integrated with any simulator. We provide an integration tutorial.
- We integrate M3SA with OpenDC, and can be enabled with flag 
-a
 when running OpenDC, via 
ScenarioCli.kt
.

Screenshot 2024-08-06 at 6 15 13 PM

## Paper
M3SA is part of the research efforts of AtLarge Research Team, towards _massivizing computer systems_. We conduct scientific research and follow the AtLarge Design Vision. We design, analyze, and engineer the M3SA system, towards performance, accuracy, universality, and usability. 

**M3SA: Exploring the Performance and Climate Impact of Datacenters by Multi-Model Simulation and Analysis**
Authors: Radu Nicolae (@Radu-Nicolae), Alexandru Iosup (@AlexandruIosup), Dante Niewenhuis (@DanteNiewenhuis)

## External Dependencies :four_leaf_clover:
N/A

## Breaking API Changes :warning:
N/A

*Simply specify none (N/A) if not applicable.*

### CODE REVIEW

File: opendc-experiments/opendc-m3sa/src/main/python/models/MultiModel.py [hunk 1]
```diff
@@ -0,0 +1,501 @@
...
+                MINUTES_IN_DAY = 1440
+                self.workload_time = len(raw) * self.user_input["samples_per_minute"] / MINUTES_IN_DAY
+
+            model = Model(raw_sim_data=raw, id=model_id, path=self.output_folder_path)
+            self.models.append(model)
+            model_id += 1
+
+        self.max_model_len = min([len(model.raw_sim_data) for model in self.models])
+
+    def compute_windowed_aggregation(self):
```
Comment (DanteNiewenhuis) [inline opendc-experiments/opendc-m3sa/src/main/python/models/MultiModel.py:223]: Pandas or NumPy should have functions that do this for you. 
There will always be more performant than what you can write in Python

File: opendc-experiments/opendc-m3sa/src/main/python/models/MultiModel.py [hunk 2]
```diff
@@ -0,0 +1,501 @@
...
+
+        :return: None
+        :side effect: Modifies each model's processed_sim_data attribute to contain aggregated data.
+        """
+        if self.plot_type != "cumulative":
+            for model in self.models:
+                numeric_values = model.raw_sim_data
+                model.processed_sim_data = self.mean_of_chunks(numeric_values, self.window_size)
+
+    def generate_plot(self):
```
Comment (DanteNiewenhuis) [inline opendc-experiments/opendc-m3sa/src/main/python/models/MultiModel.py:237]: Why would this function not be called with the plot type?
It is not clear to me where this.plot_type is set.

Comment (DanteNiewenhuis) [reply opendc-experiments/opendc-m3sa/src/main/python/models/MultiModel.py:237]: Also, I thought the idea was to automatically create multiple plots?

File: opendc-experiments/opendc-m3sa/src/main/python/models/Model.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+    Methods:
+        parse_trackr(): Reads additional configuration and metadata from a JSON file named 'trackr.json', enhancing the model with detailed context information.
+
+    Usage:
+        Model objects are typically instantiated with raw data from simulation outputs and an identifier. After instantiation,
+        the 'parse_trackr' method can be called to load additional experimental details from a corresponding JSON file.
+    """
+
+    def __init__(self, raw_sim_data, id, path):
```
Comment (DanteNiewenhuis) [inline opendc-experiments/opendc-m3sa/src/main/python/models/Model.py:39]: You can just turn this into a DataClass

Comment (DanteNiewenhuis) [reply opendc-experiments/opendc-m3sa/src/main/python/models/Model.py:39]: Given that this class seems to be purely holding data

File: opendc-experiments/opendc-m3sa/src/main/python/input_parser.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+        raise ValueError("Invalid value for meta_function. Please select between 'mean', 'median', !!!!!!!to be updated in the end!!!!!!!!.")
+
+    # raise a warning
+    if not input_json["multimodel"] and input_json["metamodel"]:
+        warnings.warn("Warning: Cannot have a Meta-Model without a Multi-Model. No computation made.")
+
+    return input_json
+
+
+def find_root_dir():
```
Comment (DanteNiewenhuis) [inline opendc-experiments/opendc-m3sa/src/main/python/input_parser.py:106]: This is fairly dangerous code. 
The chance of this failing is very high. 
For instance, what if I have multiple README.md files in my project?
What if I run m3sa from another folder?

File: opendc-experiments/opendc-m3sa/src/main/python/simulator_specifics.py [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+"""
+
+"""
+SIMULATION_DATA_FILE (str): The name of the file containing the simulation data. Enter only the name, not the path, not
+the extension. The data file must be parquet format.
+
+✅ Good: "host", "simulation_data", "cats_predictions"
+❌ Wrong: "host.json", "opendc/folder_x/folder_y/data"
+"""
+SIMULATION_DATA_FILE = "host"  # opendc outputs in file host.parquet
```
Comment (DanteNiewenhuis) [inline opendc-experiments/opendc-m3sa/src/main/python/simulator_specifics.py:14]: This file should be removed. 
Given that this is the integrated version of m3sa, it can by definition not be used by other simulators.

**Review by DanteNiewenhuis — CHANGES_REQUESTED** [review]
There are several things that need to be updated for this PR to be accepted. I have divided this review into two parts. First, I will discuss things that need to be updated for me to accept this PR. Second, I will discuss some things I strongly suggest improving, but will not stop the PR from being accepted. I do however expect you to take a look at them, especially the easy to fix comments. I have also left many comments in the files. 

**Mandatory**

1. Rename opendc-m3sa to opendc-experiment-m3sa so it adheres to the overall naming scheme of OpenDC.
- [RN] **Done**
2. There are multiple files that are changed in this PR that should not have been touched at all. For instance, all the package.json and package-lock.json files should not be changed. This is also why the docker is not building at the moment.
- [RN] **Done**
3. ExperimentCli, and Scenarios should not be changed for the functionalities m3sa currently brings. m3sa should have its own Cli that can be executed separately. m3sa should create Scenarios to run, and execute them using ScenarioRunner. 
- [RN] **Done - not changed anymore. M3SACli and ExperimentCli are now, perhaps temporary, two separate entities.**
4. It should be much clearer how m3sa currently runs. I have been going through the Python code, and it is not clear to me what the order of execution is, where the actual simulation is being executed etc.
- [RN] **Not clear to me what is the mandatory part regarding (4). I'll assume it's documentation, which is (already) done -- input guideline, readme of m3sa, and integration tutorial.**
5. Add a requirements.txt 
- [RN] **Done, you can find it in the root folder of the python code**
7. Add a README to the m3sa experiment in which you detail how to run m3sa
- [RN] **(Already) Done**

**Suggestions**

I will leave the Python files to you. However, I do have some suggestions:

1. You are executing a lot of functions in the init of your classes. For instance, MetaModel computes everything and creates all outputs. This is not common in Python, and makes it difficult to understand what is happening. 
3. Not everything has to happen inside a class. This is not Java, you can make separate functions.
4. Please take a look at Pandas and NumPy functions. There are multiple functions you have written yourself that are already implemented by either Pandas or NumPy. These will always be faster than whatever we can write in Python.
5. Adding Typing could greatly improve your Python code. Currently, users either need a good understanding of the code to understand what is expected.

File: opendc-experiments/opendc-experiments-base/build.gradle.kts
[the remark this answers is not in the document]
Comment (Radu-Nicolae) [reply opendc-experiments/opendc-experiments-base/build.gradle.kts:1]: both m3sa and experiment-base use, in one way or another, each other.

File: opendc-experiments/opendc-experiments-base/src/main/kotlin/org/opendc/experiments/base/runner/ExperimentCli.kt
[the remark this answers is not in the document]
Comment (Radu-Nicolae) [reply opendc-experiments/opendc-experiments-base/src/main/kotlin/org/opendc/experiments/base/runner/ExperimentCli.kt:1]: let's see if we have a better alternative -- how would you suggest to make this change? what code to add in which file?

[the remark this answers is not in the document]
Comment (Radu-Nicolae) [reply opendc-experiments/opendc-experiments-base/src/main/kotlin/org/opendc/experiments/base/runner/ExperimentCli.kt:1]: [TOTALK]#1

File: opendc-experiments/opendc-experiments-base/src/main/kotlin/org/opendc/experiments/base/scenario/ExperimentFactories.kt [hunk 6]
```diff
@@ -79,6 +79,7 @@ public fun getExperiment(experimentSpec: ExperimentSpec): List<Scenario> {
                 carbonTracePath = scenarioSpec.carbonTracePath,
                 exportModelSpec = scenarioSpec.exportModel,
                 outputFolder = outputFolder,
+                m3saSetup = experimentSpec.m3saSetup,
```
[the remark this answers is not in the document]
Comment (Radu-Nicolae) [reply opendc-experiments/opendc-experiments-base/src/main/kotlin/org/opendc/experiments/base/scenario/ExperimentFactories.kt:82]: yes, m3saSetup is used in scenario-post run. triggers and links the scenario with m3sa

File: opendc-experiments/opendc-experiments-base/src/main/kotlin/org/opendc/experiments/base/scenario/specs/ExperimentSpec.kt
[the remark this answers is not in the document]
Comment (Radu-Nicolae) [reply opendc-experiments/opendc-experiments-base/src/main/kotlin/org/opendc/experiments/base/scenario/specs/ExperimentSpec.kt:1]: please suggest a better alternative, what files to change and with which code

[the remark this answers is not in the document]
Comment (Radu-Nicolae) [reply opendc-experiments/opendc-experiments-base/src/main/kotlin/org/opendc/experiments/base/scenario/specs/ExperimentSpec.kt:1]: [TOTALK#2]

File: opendc-experiments/opendc-m3sa/src/main/python/input_parser.py (hunk 4 above)
Comment (Radu-Nicolae) [reply opendc-experiments/opendc-m3sa/src/main/python/input_parser.py:106]: 1. if you have multiple README.md files, in the same folder (i.e., the root folder), you break the law of operating systems :))
2. I agree this can be improved -- what better ways to ensure we are in the root folder do we have?

Comment (Radu-Nicolae) [reply opendc-experiments/opendc-m3sa/src/main/python/input_parser.py:106]: [TOTALK#3]

File: opendc-experiments/opendc-m3sa/src/main/python/models/Model.py (hunk 3 above)
Comment (Radu-Nicolae) [reply opendc-experiments/opendc-m3sa/src/main/python/models/Model.py:39]: I see. How would we benefit from this, @DanteNiewenhuis?

Comment (Radu-Nicolae) [reply opendc-experiments/opendc-m3sa/src/main/python/models/Model.py:39]: [TODO#1] read about DataClasses

File: opendc-experiments/opendc-m3sa/src/main/python/simulator_specifics.py (hunk 5 above)
Comment (Radu-Nicolae) [reply opendc-experiments/opendc-m3sa/src/main/python/simulator_specifics.py:14]: true. will remove, nice catch.

[TODO#2]

File: opendc-web/opendc-web-ui/package-lock.json
[the remark this answers is not in the document]
Comment (Radu-Nicolae) [reply opendc-web/opendc-web-ui/package-lock.json:1]: this is the result of updating and upgrading NPM. Is there any problem related to this?

File: package.json
[the remark this answers is not in the document]
Comment (Radu-Nicolae) [reply package.json:1]: [TODO#3] i'll remove it

File: settings.gradle.kts
[the remark this answers is not in the document]
Comment (Radu-Nicolae) [reply settings.gradle.kts:1]: opendc-experiment-m3sa is inaccurate and redundant. similarly to opendc-experiment-base, opendc-experiment-faas, opendc-experiment-tf20.

these should all be changed and remove experiment.
=> opendc-tf20, opendc-base, opendc-faas, opendc-m3sa.

**For the sake of consistency, I will rename to opendc-experiments-m3sa. Consider that redundant and inaccurate namings diminish the codebase readability and overall quality.**

[the remark this answers is not in the document]
Comment (Radu-Nicolae) [reply settings.gradle.kts:1]: [TODO#4]

File: site/package-lock.json
[the remark this answers is not in the document]
Comment (Radu-Nicolae) [reply site/package-lock.json:1]: idem above - see my NPM comment

File: opendc-experiments/opendc-experiments-base/build.gradle.kts
[the remark this answers is not in the document]
Comment (DanteNiewenhuis) [reply opendc-experiments/opendc-experiments-base/build.gradle.kts:1]: In which way does experiment-base use m3sa? 
Also this would be circular importing, which is not possible.

File: settings.gradle.kts
[the remark this answers is not in the document]
Comment (DanteNiewenhuis) [reply settings.gradle.kts:1]: I know that this might not be a somewhat older style of naming things (and thus redundant). 
However, that is something to solve in the future (if we want).

File: opendc-experiments/opendc-experiments-base/src/main/kotlin/org/opendc/experiments/base/scenario/ExperimentFactories.kt
[the remark this answers is not in the document]
Comment (Radu-Nicolae) [reply opendc-experiments/opendc-experiments-base/src/main/kotlin/org/opendc/experiments/base/scenario/ExperimentFactories.kt:1]: okido, will refactor

File: settings.gradle.kts
[the remark this answers is not in the document]
Comment (Radu-Nicolae) [reply settings.gradle.kts:1]: let's do that in the next PR. I'll handle.

File: opendc-experiments/opendc-m3sa/src/main/python/models/MultiModel.py (hunk 1 above)
Comment (Radu-Nicolae) [reply opendc-experiments/opendc-m3sa/src/main/python/models/MultiModel.py:223]: oh, I see. How to adapt this?

File: opendc-experiments/opendc-m3sa/src/main/python/models/MultiModel.py (hunk 2 above)
Comment (Radu-Nicolae) [reply opendc-experiments/opendc-m3sa/src/main/python/models/MultiModel.py:237]: well, not really. this depends on the user -- the plotting process takes time as well, right? Why wasting time on plotting if the user may want to only get the data of the metamodel? 

see more in the paper §4, the Meta-Model.

Comment (Radu-Nicolae) [reply opendc-experiments/opendc-m3sa/src/main/python/models/MultiModel.py:237]: the plot_type is set by the user -- either time_series, cumulative, cumulative_time_series.

For more details, have a look at NFRs for §3 and §4 of the paper for better clarification. alternatively, have a look at the documentation.

File: opendc-experiments/opendc-m3sa/src/main/python/models/Model.py (hunk 3 above)
Comment (Radu-Nicolae) [reply opendc-experiments/opendc-m3sa/src/main/python/models/Model.py]: Solved!

File: package.json
[the remark this answers is not in the document]
Comment (Radu-Nicolae) [reply package.json:1]: removed

File: opendc-experiments/opendc-experiments-base/src/main/kotlin/org/opendc/experiments/base/scenario/specs/ExperimentSpec.kt
Comment (Radu-Nicolae) [inline opendc-experiments/opendc-experiments-base/src/main/kotlin/org/opendc/experiments/base/scenario/specs/ExperimentSpec.kt:1]: m3saSetup in ExperimentSpec -- non mandatory to setup, but part of an experiment, used when parsing the JSON file. this is part of the JSON file parser.

File: opendc-web/opendc-web-ui/package-lock.json
Comment (DanteNiewenhuis) [inline opendc-web/opendc-web-ui/package-lock.json:1]: Remove this change

File: opendc-experiments/opendc-experiments-m3sa/src/main/kotlin/org/opendc/experiments/m3sa/runner/M3SARunner.kt
Comment (DanteNiewenhuis) [inline opendc-experiments/opendc-experiments-m3sa/src/main/kotlin/org/opendc/experiments/m3sa/runner/M3SARunner.kt:1]: As far as I can see, this file is not needed. You can just import runExperiment from experiment-base

Comment (DanteNiewenhuis) [reply opendc-experiments/opendc-experiments-m3sa/src/main/kotlin/org/opendc/experiments/m3sa/runner/M3SARunner.kt:1]: If you want to do something different here at some point, you can create it later.

File: opendc-experiments/opendc-experiments-m3sa/src/main/kotlin/org/opendc/experiments/m3sa/scenario/M3SAFactories.kt
Comment (DanteNiewenhuis) [inline opendc-experiments/opendc-experiments-m3sa/src/main/kotlin/org/opendc/experiments/m3sa/scenario/M3SAFactories.kt:1]: Why does this have to be a different file than just the ExperimentFactory?

**Review by DanteNiewenhuis — CHANGES_REQUESTED** [review]
The PR is better than the previous version, but I still have some issues.

First, you are still changing things in the ExperimentSpecs. This should not be needed. 

Second, As far as I can see, the Python files have not been changed. This means that I will not understand what is happening. 

Finally, it seems like there is no pre-processing being done. As far as I understand, the multi/meta model creates multiple models from the given input. I cannot find any such pre processing in the code at the moment.

File: site/docs/documentation/Input/Scenario.md [hunk 7]
```diff
@@ -8,18 +8,19 @@ In the following section, we describe the different components of the schema.
...
+| topologies         | List[[Topology](#topology)]                  | yes       | N/A      | List of topologies used in the scenario.                          |
+| workloads          | List[[Workload](#workload)]                  | yes       | N/A      | List of workloads to be executed within the scenario.             |
+| allocationPolicies | List[[AllocationPolicy](#allocation-policy)] | yes       | N/A      | Allocation policies used for resource management in the scenario. |
+| failureModels      | List[[FailureModel](#failuremodel)]          | no        | empty    | List of failure models to simulate various types of failures.     |
+| exportModels       | List[[ExportModel](#exportmodel)]            | no        | empty    | Specifications for exporting data from the simulation.            |
+| carbonTracePaths   | List[string]                                 | no        | null     | Paths to carbon footprint trace files.                            |
+| m3saSetup          | string                                       | no        | null     | Path to the cofigurator file of M3SA.                             |
```
Comment (DanteNiewenhuis) [inline site/docs/documentation/Input/Scenario.md:20]: This needs to be removed

Comment (Radu-Nicolae) [reply site/docs/documentation/Input/Scenario.md:20]: ok, I'll revert it to the current scenario, from the master branch.

File: opendc-web/opendc-web-ui/package-lock.json
Comment (DanteNiewenhuis) [inline opendc-web/opendc-web-ui/package-lock.json:1]: These versions should not be changed, Could you please update this and push again?

File: site/package-lock.json
Comment (DanteNiewenhuis) [inline site/package-lock.json:1]: Could you remove this file from your push?

**Review by DanteNiewenhuis — APPROVED** [review]
